### PR TITLE
Multiwallet

### PR DESCRIPTION
**Pull request by brunobar79**
This PR adds experimental support for multiple wallets, which can be imported via:
- Seed
- Mnemonic (For. ex  MetaMask or any other compatible wallet)
- Private Key
- Ethereum Address (watch only mode) 
- ENS Name  (watch only mode)

The watch only mode allows users to import ANY account and having an almost fully functional wallet without any security risk.  Could be used to monitor ledger or  trezor accounts,  etc.

Also while importing from seed or mnemonic, if there's more than 1 account account that have been used (has at least 1 transaction) it will be enabled automatically

You can  to edit the name & color of each wallet & account & also delete it  (if it's not currently selected one).

**Some other notes:**
- This PR includes a keychain migration
- The UI code  will change very soon since designs are in progress, but this is ready to be merged with the experimental flag turned off.  

**For reference, this is all the scenarios that I've tested against:**

Starting from develop (To be able to test migrations):

- [x]  New empty wallet,  kill the app, switch branches, restart the app
- [x]  Import wallet via pkey,  kill the app, switch branches, restart the app
- [x]  Import wallet via seed,  kill the app, switch branches, restart the app

From this  branch:

- [x] Brand new*
- [x] Brand new* => import wallet via pkey (no account creation)
- [x] Brand new* => import wallet via seed using dev wallet (should autodiscover 2 accounts by default)
- [x] After importing via seed,  create account  switch accounts in the same wallet
- [x] Import  another wallet via seed - switch accounts in diff wallets
- [x] Import via ENS (readonly) - try to execute non read only stuff (send, swap, savings, etc)
- [x] Import via 0x (readonly) - try to execute non read only stuff (send, swap, savings, etc)
- [x] Switch from read only to non read only stuff and execute accounts
- [x] Edit { Read Only | Private Key | Walllet | Account } name & color => Delete

*Brand new means: wipe simulator (Device => Erase all content and settings) and reinstall  the app to reproduce the first time user scenario

### CODE REVIEW

File: src/config/experimental.js [hunk 1]
```diff
@@ -9,3 +9,4 @@ export const chartExpandedAvailable = false;
 export const nativeTransactionListAvailable = true;
 export const nativeButtonPressAnimationAvailable = true;
 export const isAvatarPickerAvailable = false;
+export const isMultiwalletAvailable = true;
```
Comment (jinchung) [inline src/config/experimental.js:12]: let's not merge in this as true just yet!

File: src/screens/Routes/helpers.js [hunk 2]
```diff
@@ -42,6 +42,10 @@ export const expandedPresetWithTransitions = presetWithTransition(
   expandedPreset
 );
 
+export const expandedReversePresetWithTransitions = presetWithTransition(
+  expandedPreset
```
Comment (jinchung) [inline src/screens/Routes/helpers.js:46]: Did you mean for this to be "expandedPresetReverse" instead of "expandedPreset"

**Comment by jinchung** [comment]
Things to remember (these are non-merge-blocking):
- we'll need to update the avatar builder logic
- remove account name / color from useAccountSettings

File: src/model/wallet.js [hunk 3]
```diff
@@ -259,37 +439,125 @@ const saveWalletDetails = async (seedPhrase, privateKey, address) => {
...
 };
 
-const loadPrivateKey = async (
+export const getAllWallets = async (
   authenticationPrompt = lang.t('wallet.authenticate.please')
 ) => {
   try {
-    return keychain.loadString(privateKeyKey, {
+    return keychain.loadObject(allWalletsKey, {
       authenticationPrompt,
```
Comment (jinchung) [inline src/model/wallet.js:561]: This does not need the authenticationPrompt

File: src/model/wallet.js [hunk 4]
```diff
@@ -259,37 +439,125 @@ const saveWalletDetails = async (seedPhrase, privateKey, address) => {
...
 };
 
-const saveSeedPhrase = async (seedPhrase, accessControlOptions = {}) => {
-  await keychain.saveString(seedPhraseKey, seedPhrase, accessControlOptions);
+export const getSelectedWallet = async (
+  authenticationPrompt = lang.t('wallet.authenticate.please')
+) => {
+  try {
+    return keychain.loadObject(selectedWalletKey, {
+      authenticationPrompt,
```
Comment (jinchung) [inline src/model/wallet.js:535]: authentication prompt not needed here

File: src/model/keychain.js [hunk 5]
```diff
@@ -32,9 +32,9 @@ export async function loadString(key, authenticationPrompt) {
...
 }
 
-export async function saveObject(key, value) {
+export async function saveObject(key, value, accessControlOptions) {
   const jsonValue = JSON.stringify(value);
-  await saveString(key, jsonValue);
+  await saveString(key, jsonValue, accessControlOptions);
 }
 
 export async function loadObject(key) {
```
Comment (jinchung) [inline src/model/keychain.js:40]: this function should include the authentication prompt parameter and pass it along to loadString

File: src/redux/wallets.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    index: newIndex,
+    label: '',
+    visible: true,
+  });
+
+  // Save all the wallets
+  saveAllWallets(wallets);
+  // Set the address selected (KEYCHAIN)
+  await saveAddress(account.address);
+  // Set the wallet selected (REDUX)
```
Comment (jinchung) [inline src/redux/wallets.js:73]: comment should say `Set the wallet selected (KEYCHAIN)`

File: src/redux/wallets.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+export const createAccountForWallet = id => async (dispatch, getState) => {
+  const { wallets } = getState().wallets;
+  let index = 0;
+  wallets[id].addresses.forEach(
+    account => (index = Math.max(index, account.index))
+  );
+  const newIndex = index + 1;
+  const account = await generateAccount(id, newIndex);
+  wallets[id].addresses.push({
```
Comment (jinchung) [inline src/redux/wallets.js:61]: Can we do this update on the state object in an immutable fashion

File: src/model/wallet.js [hunk 8]
```diff
@@ -56,11 +77,6 @@ export const loadWallet = async () => {
   return null;
 };
 
-export const getChainId = async () => {
-  const wallet = await loadWallet();
-  return get(wallet, 'provider.chainId');
-};
-
 export const createTransaction = async (
```
Comment (jinchung) [inline src/model/wallet.js:80]: This function seems no longer used - we can remove it

File: src/model/wallet.js [hunk 9]
```diff
@@ -222,24 +238,187 @@ export const loadAddress = async () => {
...
+    // Get all wallets
+    const allWalletsResult = await getAllWallets();
+    const allWallets = get(allWalletsResult, 'wallets', {});
+
+    if (isImported) {
+      // Checking if the generated account already exists
+      const alreadyExisting = Object.keys(allWallets).some(key => {
+        const someWallet = allWallets[key];
+        return someWallet.addresses.some(account => {
+          return account.address === wallet.address && account.type === type;
```
Comment (jinchung) [inline src/model/wallet.js:333]: are the addresses from ethers wallet object always checksummed?

File: src/model/wallet.js [hunk 10]
```diff
@@ -222,24 +238,187 @@ export const loadAddress = async () => {
...
 
-const createWallet = async seed => {
+const loadPrivateKey = async (
+  authenticationPrompt = lang.t('wallet.authenticate.please')
+) => {
+  try {
+    const isSeedPhraseMigrated = await keychain.loadString(
+      seedPhraseMigratedKey,
+      {
+        authenticationPrompt,
```
Comment (jinchung) [inline src/model/wallet.js:248]: No need for authentication prompt here

File: src/model/wallet.js [hunk 11]
```diff
@@ -298,6 +566,124 @@ const loadPrivateKey = async (
...
+
+export const loadSeedPhraseAndMigrateIfNeeded = async (
+  id,
+  authenticationPrompt = lang.t('wallet.authenticate.please_seed_phrase')
+) => {
+  try {
+    const isSeedPhraseMigrated = await keychain.loadString(
+      seedPhraseMigratedKey,
+      {
+        authenticationPrompt,
```
Comment (jinchung) [inline src/model/wallet.js:671]: authenticationPrompt not needed here

File: src/model/wallet.js [hunk 12]
```diff
@@ -222,24 +238,187 @@ export const loadAddress = async () => {
...
+        } else {
+          lookup = false;
+        }
+      }
+    }
+
+    // if imported and we have only one account, we name it.
+    // If we have more than one account, we name the wallet
+    let walletName = DEFAULT_WALLET_NAME;
+    if (isImported) {
```
Comment (jinchung) [inline src/model/wallet.js:401]: perhaps move name check to here (isImported && name)

File: src/screens/ImportSeedPhraseSheet.js [hunk 13]
```diff
@@ -80,11 +82,30 @@ const StyledInput = styled(Input)`
...
 `;
 
-const ConfirmImportAlert = onSuccess =>
+const ConfirmImportAlert = (onSuccess, navigate) =>
   Alert({
     buttons: [
       {
-        onPress: onSuccess,
+        onPress: () =>
+          navigate('OverlayExpandedAssetScreen', {
```
Comment (jinchung) [inline src/screens/ImportSeedPhraseSheet.js:90]: Routes from routeNames for this

File: src/components/expanded-state/WalletProfileCreator.js [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+        </AssetPanel>
+      </FloatingPanels>
+    </KeyboardFixedOpenLayout>
+  );
+}
+
+WalletProfileCreator.propTypes = {
+  actionType: PropTypes.string,
+  address: PropTypes.string,
+  color: PropTypes.number,
```
Comment (jinchung) [inline src/components/expanded-state/WalletProfileCreator.js:233]: color propType not needed

File: src/components/expanded-state/WalletProfileCreator.js [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+
+export default function WalletProfileCreator({
+  actionType,
+  address,
+  isCurrentProfile,
+  isNewProfile,
+  isDeletable,
+  onCloseModal,
+  profile,
+  setIsLoading,
```
Comment (jinchung) [inline src/components/expanded-state/WalletProfileCreator.js:40]: doesn't seem setIsLoading is passed in from anywhere?

**Comment by jinchung** [comment]
If I import a wallet, and in the modal to add its name I hit cancel, it will still import the wallet.

**Comment by jinchung** [comment]
I'm able to add the same wallet and type multiple times

File: src/screens/ChangeWalletModal.js [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+      let isDeletable = false;
+      if (Object.keys(wallet).length > 1 && selectedWallet.id !== id) {
+        isDeletable = true;
+      }
+      navigate('ExpandedAssetScreen', {
+        actionType: 'Import',
+        address: undefined,
+        asset: [],
+        isDeletable,
+        onCloseModal: async args => {
```
Comment (jinchung) [inline src/screens/ChangeWalletModal.js:83]: onCloseModal is called by the cancel function in WalletProfileCreator without any args, which still triggers the onSuccess to toggle the importing state

File: src/screens/ChangeWalletModal.js [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+const ChangeWalletModal = () => {
+  const { wallets, selected: selectedWallet } = useWallets();
+
+  const { goBack, navigate } = useNavigation();
+  const dispatch = useDispatch();
+  const { accountAddress } = useAccountSettings();
+  const initializeWallet = useInitializeWallet();
+  let rowsCount = 0;
+  if (wallets) {
+    Object.keys(wallets).forEach(key => {
```
Comment (jinchung) [inline src/screens/ChangeWalletModal.js:28]: can use useMemo for rowsCount

File: src/screens/ChangeWalletModal.js [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+  const onEditWallet = useCallback(
+    id => {
+      const wallet = wallets[id];
+      if (!selectedWallet) return;
+
+      let isDeletable = false;
+      if (Object.keys(wallet).length > 1 && selectedWallet.id !== id) {
+        isDeletable = true;
+      }
+      navigate('ExpandedAssetScreen', {
```
Comment (jinchung) [inline src/screens/ChangeWalletModal.js:78]: use Route names

File: src/screens/ChangeWalletModal.js [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+    (wallet_id, address) => {
+      const wallet = wallets[wallet_id];
+      const account = wallet.addresses.find(
+        account => account.address === address
+      );
+      let isDeletable = false;
+      if (accountAddress !== address) {
+        isDeletable = true;
+      }
+      navigate('ExpandedAssetScreen', {
```
Comment (jinchung) [inline src/screens/ChangeWalletModal.js:115]: use Route names

File: src/screens/ChangeWalletModal.js [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+      } catch (e) {
+        logger.log('Error while trying to add account', e);
+      }
+    },
+    [dispatch, initializeWallet]
+  );
+
+  const onPressImportSeedPhrase = useCallback(() => {
+    goBack();
+    navigate('ImportSeedPhraseSheet');
```
Comment (jinchung) [inline src/screens/ChangeWalletModal.js:174]: Use route names

File: src/screens/ChangeWalletModal.js [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+    listHeight = 298;
+  }
+
+  const onChangeAccount = useCallback(
+    async (wallet_id, address) => {
+      try {
+        const wallet = wallets[wallet_id];
+        dispatch(walletsSetSelected(wallet));
+        dispatch(addressSetSelected(address));
+        await initializeWallet();
```
Comment (jinchung) [inline src/screens/ChangeWalletModal.js:60]: Currently, switching back and forth between accounts will create more Zerion connections (with the old ones still listening to their messages too.) If you go from wallet A to wallet B back to wallet A, you'll end up with 3 websockets, 2 for A, 1 for B. We can discuss how we want to manage these ideally. We could share the same websocket and have it subscribe to multiple addresses, we'll just need to change how we handle subscribe/unsubscribe and the different redux states for the different addresses.

File: src/model/wallet.js [hunk 22]
```diff
@@ -298,6 +566,124 @@ const loadPrivateKey = async (
...
+    logger.log('Error generating account for keychain', id, error);
+  }
+};
+
+export const migrateSecrets = async () => {
+  try {
+    const publicAccessControlOptions = {
+      accessible: ACCESSIBLE.ALWAYS_THIS_DEVICE_ONLY,
+    };
+    const seedPhrase = await loadSeedPhrase();
```
Comment (jinchung) [inline src/model/wallet.js:618]: can we add something to the function name `loadSeedPhrase` that makes it obvious that it was part of the old version

**Comment by jinchung** [comment]
Not sure if this has already been decided, but just wanted to check what we would be doing with deleted wallets / accounts private keys on keychain.

File: src/screens/ChangeWalletModal.js [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+    id => {
+      const wallet = wallets[id];
+      if (!selectedWallet) return;
+
+      let isDeletable = false;
+      if (Object.keys(wallet).length > 1 && selectedWallet.id !== id) {
+        isDeletable = true;
+      }
+      navigate('ExpandedAssetScreen', {
+        actionType: 'Import',
```
Comment (jinchung) [inline src/screens/ChangeWalletModal.js:79]: actionType not needed here

File: src/screens/ChangeWalletModal.js [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+      const wallet = wallets[wallet_id];
+      const account = wallet.addresses.find(
+        account => account.address === address
+      );
+      let isDeletable = false;
+      if (accountAddress !== address) {
+        isDeletable = true;
+      }
+      navigate('ExpandedAssetScreen', {
+        actionType: 'Import',
```
Comment (jinchung) [inline src/screens/ChangeWalletModal.js:116]: actionType not needed here

**Comment by jinchung** [comment]
Also, I had tried adding an invalid ENS name, and it let me go through with importing, but ended up not doing anything (though there was no error telling me the ENS I was trying to add wasn't actually valid)

File: src/model/wallet.js (hunk 9 above)
Comment (brunobar79) [reply src/model/wallet.js:333]: Yes

Comment (brunobar79) [reply src/model/wallet.js:333]: I'll add it anyway cause read only wallets don't go through ethers

**Comment by brunobar79** [comment]
> If I import a wallet, and in the modal to add its name I hit cancel, it will still import the wallet.

The cancel button is not to prevent importing a wallet but to skip naming.  I suggest we revisit it once the designs are ready

File: src/screens/ChangeWalletModal.js (hunk 16 above)
Comment (brunobar79) [reply src/screens/ChangeWalletModal.js:83]: This is related to my response here: https://github.com/rainbow-me/rainbow/pull/606#issuecomment-624941336

File: src/model/wallet.js (hunk 3 above)
Comment (brunobar79) [reply src/model/wallet.js:538]: We decided to add the param to loadObject

File: src/model/wallet.js (hunk 4 above)
Comment (brunobar79) [reply src/model/wallet.js:516]: We decided to add the param to loadObject

File: src/model/wallet.js [hunk 25]
```diff
@@ -222,24 +238,187 @@ export const loadAddress = async () => {
...
+      return privateKey;
+    }
+  } catch (error) {
+    captureException(error);
+    return null;
+  }
+};
+
+export const saveAddress = async (address, accessControlOptions = {}) => {
+  await keychain.saveString(addressKey, address, accessControlOptions);
```
[the remark this answers is not in the document]
Comment (brunobar79) [reply src/model/wallet.js:254]: that's a good idea. ended up exporting it since it's also used in migrations

**Comment by jinchung** [comment]
> > If I import a wallet, and in the modal to add its name I hit cancel, it will still import the wallet.
> 
> The cancel button is not to prevent importing a wallet but to skip naming. I suggest we revisit it once the designs are ready

Gotcha - thanks!

File: src/redux/wallets.js [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+    payload: wallet,
+    type: WALLETS_SET_SELECTED,
+  });
+};
+export const addressSetSelected = address => () => {
+  saveAddress(address, publicAccessControlOptions);
+};
+
+export const createAccountForWallet = id => async (dispatch, getState) => {
+  const { wallets } = { ...getState().wallets };
```
Comment (jinchung) [inline src/redux/wallets.js:56]: This line doesn't need the destructuring as it is taken care of in the next line!

Comment (brunobar79) [reply src/redux/wallets.js:56]: oops! forgot to change that

**Comment by brunobar79** [comment]
TODO before merging:
- Take into account avatars
- One more round of testing

**Pull request merged**